src/dd_header.F90

```diff
@@ -0,0 +1,110 @@
+    integer :: neighbor_meshbins(42) ! domain meshbins
+    type(DictIntInt) :: bins_dict    ! reverse of neighbor_meshbins
+
+    ! DD communication information
+    integer :: comm   ! communicator for fission bank and tallies in this domain
+    integer :: comm_domain_masters ! communicator for local domain masters
+    integer :: rank   ! this processes's rank in domain sub-communicator
+    logical :: local_master         ! master process in this domain?
+    integer :: n_domain_procs       ! number of processors in domain sub-comm
+    integer :: max_domain_procs     ! max number of procs in any other sub-comm
+
+    ! User-specified ditribution of load
+    real(8), allocatable :: domain_load_dist(:) ! index is domain meshbin
+
+    ! Calculated number of procs per domain
+    integer, allocatable :: domain_n_procs(:) ! index is domain meshbin
+
+    ! Global ranks of the local masters in each domain, used for computing send
+    ! ranks
+    integer, allocatable :: domain_masters(:) ! index is domain meshbin
+
+    ! For non-DD runs we re-ruse the same particle data structure for each
+    ! history, but for DD runs we need to store particles if they would be
+    ! transmitted to a neighboring domain.
+    type(Particle), allocatable :: particle_buffer(:)
+    integer(8) :: size_particle_buffer ! size of particle_buffer
+
+    ! During simulation we keep track of where particles will scatter out to
+    ! with p % outscatter_destination, and then send them later during
+    ! syncronize_banks
+    type(ParticleBuffer), allocatable :: send_buffer(:)
+    type(ParticleBuffer), allocatable :: recv_buffer(:)
+    integer(8) :: size_send_buffer ! size of buffer, resized if needed
+    integer(8) :: size_recv_buffer ! size of buffer, resized if needed
+
+    ! Scatter information
+    integer              :: n_global_scatters ! # of scatters in last DD stage
+    integer, allocatable :: send_rank_info(:) ! # of particles to send to bin
+    integer, allocatable :: recv_rank_info(:) ! # of particles to recv from bin
+    integer              :: n_inscatt         ! # of particles received
+    integer, allocatable :: proc_finish(:)    ! finish idx in scatt array
+
+    ! Before synchronizing scatters, all processors need to know how many
+    ! particles are going everywhere
+    integer, allocatable :: n_scatters_neighborhood(:,:) ! (from_bin, to_bin)
+    integer, allocatable :: n_scatters_domain(:) ! (to_bin) just this domain
+    integer, allocatable :: n_scatters_local(:) ! (to bin) just this processor
+    integer, allocatable :: scatter_offest(:) ! (to bin) offset this processor
+
+    ! To help with setting nodemaps, we can run in a mode that counts all
+    ! particle interactions in a domain(a counter in the while(alive) particle
+    ! loop.  This will be printed as output for the domain masters only
+    logical    :: count_interactions = .false.
+    integer(8) :: interaction_count = 0_8
+
+  end type DomainDecomType
```

Comment:
@smharper If `initialize_domain_decomp` is made a type-bound procedure, `'use global'` is needed in `module dd_header`, but `'use dd_header'` exists in `module dd_header`, causing compiling failure `'Circular dependency'`. What's your suggestion to this?